### PR TITLE
core: revise logic for determining screenshot dimensions

### LINKED ISSUE #14824: Incorrect width and height of full page screenshot
### FAQ

- [X] Yes, my issue is not about [variability](https://github.com/GoogleChrome/lighthouse/blob/main/docs/variability.md) or [throttling](https://github.com/GoogleChrome/lighthouse/blob/main/docs/throttling.md).
- [X] Yes, my issue is not about a specific accessibility audit (file with [axe-core](https://github.com/dequelabs/axe-core) instead).

### URL

https://wp-rocket.me

### What happened?

I generated a JSON report for this URL: https://wp-rocket.me

From the JSON report, I extracted the webp image buffer and converted it to a file. 

While I was checking the file attributes, I noticed the width and height mentioned in the report are different compared to what the actual size is.

Actual Size: 412 × 8795 pixels
Size mentioned in report: 417 x 8902 pixels

### What did you expect?

I expect the full-page screenshot image buffer to be the same size as mentioned in the report.

### What have you tried?

I found this issue only for this particular URL. All other URLs I tried are working fine.
Also, I'm getting the correct width and height with lighthouse 9.6.6.

### How were you running Lighthouse?

CLI, node

### Lighthouse Version

10.0.1

### Chrome Version

109.0.5414.119 (Official Build) (arm64)

### Node Version

16.18.1

### OS

Mac

### Relevant log output

```shell
$ lighthouse https://wp-rocket.me/ --output json --output-path ./test.json
  LH:ChromeLauncher Waiting for browser. +0ms
  LH:ChromeLauncher Waiting for browser... +0ms
  LH:ChromeLauncher Waiting for browser..... +503ms
  LH:ChromeLauncher Waiting for browser.....✓ +1ms
  LH:status Connecting to browser +429ms
  LH:status Navigating to about:blank +5ms
  LH:status Benchmarking machine +14ms
  LH:status Preparing target for navigation mode +1s
  LH:status Navigating to about:blank +7ms
  LH:status Preparing target for navigation +6ms
  LH:status Cleaning origin data +21ms
  LH:status Cleaning browser cache +13ms
  LH:status Preparing network conditions +18ms
  LH:status Navigating to https://wp-rocket.me/ +39ms
  LH:artifacts:getArtifact DevtoolsLog +4s
  LH:artifacts:getArtifact Trace +0ms
  LH:artifacts:getArtifact DevtoolsLog +1ms
  LH:artifacts:getArtifact Trace +0ms
  LH:artifacts:getArtifact Accessibility +0ms
  LH:artifacts:getArtifact AnchorElements +229ms
  LH:artifacts:getArtifact ConsoleMessages +19ms
  LH:artifacts:getArtifact CSSUsage +1ms
  LH:artifacts:getArtifact Doctype +17ms
  LH:artifacts:getArtifact DOMStats +1ms
  LH:artifacts:getArtifact EmbeddedContent +3ms
  LH:artifacts:getArtifact FontSize +1ms
  LH:artifacts:getArtifact Inputs +17ms
  LH:artifacts:getArtifact GlobalListeners +2ms
  LH:artifacts:getArtifact ImageElements +1ms
  LH:artifacts:getArtifact InstallabilityErrors +223ms
  LH:status Get webapp installability errors +0ms
  LH:artifacts:getArtifact InspectorIssues +0ms
  LH:artifacts:getArtifact JsUsage +1ms
  LH:artifacts:getArtifact LinkElements +0ms
  LH:artifacts:getArtifact MainDocumentContent +3ms
  LH:artifacts:getArtifact MetaElements +2ms
  LH:artifacts:getArtifact NetworkUserAgent +3ms
  LH:artifacts:getArtifact OptimizedImages +0ms
  LH:artifacts:getArtifact ResponseCompression +0ms
  LH:artifacts:getArtifact RobotsTxt +0ms
  LH:artifacts:getArtifact ServiceWorker +184ms
  LH:artifacts:getArtifact Scripts +3ms
  LH:artifacts:getArtifact SourceMaps +0ms
  LH:artifacts:getArtifact Stacks +0ms
  LH:status Collect stacks +0ms
  LH:artifacts:getArtifact TagsBlockingFirstPaint +14ms
  LH:artifacts:getArtifact TapTargets +1ms
  LH:artifacts:getArtifact TraceElements +12ms
  LH:artifacts:getArtifact ViewportDimensions +34ms
  LH:artifacts:getArtifact WebAppManifest +1ms
  LH:status Get webapp manifest +0ms
  LH:artifacts:getArtifact devtoolsLogs +1ms
  LH:artifacts:getArtifact traces +1ms
  LH:artifacts:getArtifact FullPageScreenshot +0ms
  LH:artifacts:getArtifact BFCacheFailures +1s
  LH:status Analyzing and running audits... +256ms
  LH:status Auditing: Uses HTTPS +1ms
  LH:status Auditing: Registers a service worker that controls page and `start_url` +0ms
  LH:status Auditing: Has a `<meta name="viewport">` tag with `width` or `initial-scale` +0ms
  LH:status Auditing: First Contentful Paint +1ms
  LH:status Auditing: Largest Contentful Paint +6ms
  LH:status Auditing: First Meaningful Paint +2ms
  LH:status Auditing: Speed Index +1ms
  LH:status Auditing: Screenshot Thumbnails +163ms
  LH:status Auditing: Final Screenshot +1ms
  LH:status Auditing: Total Blocking Time +0ms
  LH:status Auditing: Max Potential First Input Delay +2ms
  LH:status Auditing: Cumulative Layout Shift +2ms
  LH:status Auditing: No browser errors logged to the console +1ms
  LH:status Auditing: Initial server response time was short +0ms
  LH:status Auditing: Time to Interactive +0ms
  LH:status Auditing: User Timing marks and measures +1ms
  LH:status Auditing: Avoid chaining critical requests +0ms
  LH:status Auditing: Avoid multiple page redirects +1ms
  LH:status Auditing: Web app manifest and service worker meet the installability requirements +0ms
  LH:status Auditing: Configured for a custom splash screen +1ms
  LH:status Auditing: Sets a theme color for the address bar. +1ms
  LH:status Auditing: Manifest has a maskable icon +0ms
  LH:status Auditing: Content is sized correctly for the viewport +0ms
  LH:status Auditing: Displays images with correct aspect ratio +1ms
  LH:status Auditing: Serves images with appropriate resolution +3ms
  LH:status Auditing: Fonts with `font-display: optional` are preloaded +0ms
  LH:status Auditing: Avoids deprecated APIs +0ms
  LH:status Auditing: Minimizes main-thread work +1ms
  LH:status Auditing: JavaScript execution time +3ms
  LH:status Auditing: Preload key requests +1ms
  LH:status Auditing: Preconnect to required origins +0ms
  LH:status Auditing: All text remains visible during webfont loads +0ms
  LH:status Auditing: Diagnostics +1ms
  LH:status Auditing: Network Requests +0ms
  LH:status Auditing: Network Round Trip Times +1ms
  LH:status Auditing: Server Backend Latencies +0ms
  LH:status Auditing: Tasks +0ms
  LH:status Auditing: Metrics +0ms
  LH:status Auditing: Performance budget +1ms
  LH:status Auditing: Timing budget +1ms
  LH:status Auditing: Keep request counts low and transfer sizes small +0ms
  LH:status Auditing: Minimize third-party usage +0ms
  LH:status Auditing: Lazy load third-party resources with facades +1ms
  LH:status Auditing: Largest Contentful Paint element +0ms
  LH:status Auditing: Largest Contentful Paint image was not lazily loaded +1ms
  LH:status Auditing: Avoid large layout shifts +0ms
  LH:status Auditing: Avoid long main-thread tasks +0ms
  LH:status Auditing: Avoids `unload` event listeners +1ms
  LH:status Auditing: Avoid non-composited animations +0ms
  LH:status Auditing: Image elements have explicit `width` and `height` +0ms
  LH:status Auditing: Page has valid source maps +1ms
  LH:status Auditing: Preload Largest Contentful Paint image +0ms
  LH:status Auditing: Ensure CSP is effective against XSS attacks +0ms
  LH:status Auditing: Script Treemap Data +1ms
  LH:status Auditing: Site works cross-browser +0ms
  LH:status Auditing: Page transitions don't feel like they block on the network +1ms
  LH:status Auditing: Each page has a URL +0ms
  LH:status Auditing: `[accesskey]` values are unique +0ms
  LH:status Auditing: `[aria-*]` attributes match their roles +0ms
  LH:status Auditing: `button`, `link`, and `menuitem` elements have accessible names +0ms
  LH:status Auditing: `[aria-hidden="true"]` is not present on the document `<body>` +1ms
  LH:status Auditing: `[aria-hidden="true"]` elements do not contain focusable descendents +0ms
  LH:status Auditing: ARIA input fields have accessible names +0ms
  LH:status Auditing: ARIA `meter` elements have accessible names +0ms
  LH:status Auditing: ARIA `progressbar` elements have accessible names +0ms
  LH:status Auditing: `[role]`s have all required `[aria-*]` attributes +0ms
  LH:status Auditing: Elements with an ARIA `[role]` that require children to contain a specific `[role]` have all required children. +0ms
  LH:status Auditing: `[role]`s are contained by their required parent element +0ms
  LH:status Auditing: `[role]` values are valid +1ms
  LH:status Auditing: ARIA toggle fields have accessible names +0ms
  LH:status Auditing: ARIA `tooltip` elements have accessible names +0ms
  LH:status Auditing: ARIA `treeitem` elements have accessible names +0ms
  LH:status Auditing: `[aria-*]` attributes have valid values +0ms
  LH:status Auditing: `[aria-*]` attributes are valid and not misspelled +0ms
  LH:status Auditing: Buttons have an accessible name +0ms
  LH:status Auditing: The page contains a heading, skip link, or landmark region +1ms
  LH:status Auditing: Background and foreground colors have a sufficient contrast ratio +0ms
  LH:status Auditing: `<dl>`'s contain only properly-ordered `<dt>` and `<dd>` groups, `<script>`, `<template>` or `<div>` elements. +0ms
  LH:status Auditing: Definition list items are wrapped in `<dl>` elements +0ms
  LH:status Auditing: Document has a `<title>` element +0ms
  LH:status Auditing: `[id]` attributes on active, focusable elements are unique +0ms
  LH:status Auditing: ARIA IDs are unique +0ms
  LH:status Auditing: No form fields have multiple labels +1ms
  LH:status Auditing: `<frame>` or `<iframe>` elements have a title +0ms
  LH:status Auditing: Heading elements appear in a sequentially-descending order +0ms
  LH:status Auditing: `<html>` element has a `[lang]` attribute +0ms
  LH:status Auditing: `<html>` element has a valid value for its `[lang]` attribute +0ms
  LH:status Auditing: Image elements have `[alt]` attributes +0ms
  LH:status Auditing: `<input type="image">` elements have `[alt]` text +0ms
  LH:status Auditing: Form elements have associated labels +0ms
  LH:status Auditing: Links have a discernible name +0ms
  LH:status Auditing: Lists contain only `<li>` elements and script supporting elements (`<script>` and `<template>`). +1ms
  LH:status Auditing: List items (`<li>`) are contained within `<ul>`, `<ol>` or `<menu>` parent elements +0ms
  LH:status Auditing: The document does not use `<meta http-equiv="refresh">` +0ms
  LH:status Auditing: `[user-scalable="no"]` is not used in the `<meta name="viewport">` element and the `[maximum-scale]` attribute is not less than 5. +0ms
  LH:status Auditing: `<object>` elements have alternate text +0ms
  LH:status Auditing: No element has a `[tabindex]` value greater than 0 +0ms
  LH:status Auditing: Cells in a `<table>` element that use the `[headers]` attribute refer to table cells within the same table. +1ms
  LH:status Auditing: `<th>` elements and elements with `[role="columnheader"/"rowheader"]` have data cells they describe. +0ms
  LH:status Auditing: `[lang]` attributes have a valid value +0ms
  LH:status Auditing: `<video>` elements contain a `<track>` element with `[kind="captions"]` +0ms
  LH:status Auditing: Custom controls have associated labels +0ms
  LH:status Auditing: Custom controls have ARIA roles +0ms
  LH:status Auditing: User focus is not accidentally trapped in a region +0ms
  LH:status Auditing: Interactive controls are keyboard focusable +0ms
  LH:status Auditing: Interactive elements indicate their purpose and state +0ms
  LH:status Auditing: The page has a logical tab order +0ms
  LH:status Auditing: The user's focus is directed to new content added to the page +0ms
  LH:status Auditing: Offscreen content is hidden from assistive technology +0ms
  LH:status Auditing: HTML5 landmark elements are used to improve navigation +0ms
  LH:status Auditing: Visual order on the page follows DOM order +0ms
  LH:status Auditing: Uses efficient cache policy on static assets +0ms
  LH:status Auditing: Avoids enormous network payloads +1ms
  LH:status Auditing: Defer offscreen images +0ms
  LH:status Auditing: Eliminate render-blocking resources +1ms
  LH:status Auditing: Minify CSS +1ms
  LH:status Auditing: Minify JavaScript +5ms
  LH:status Auditing: Reduce unused CSS +2ms
  LH:status Auditing: Reduce unused JavaScript +1ms
  LH:status Auditing: Serve images in next-gen formats +1ms
  LH:status Auditing: Efficiently encode images +1ms
  LH:status Auditing: Enable text compression +1ms
  LH:status Auditing: Properly size images +1ms
  LH:status Auditing: Use video formats for animated content +3ms
  LH:status Auditing: Remove duplicate modules in JavaScript bundles +1ms
  LH:status Auditing: Avoid serving legacy JavaScript to modern browsers +1ms
  LH:status Auditing: Page has the HTML doctype +8ms
  LH:status Auditing: Properly defines charset +0ms
  LH:status Auditing: Avoids an excessive DOM size +0ms
  LH:status Auditing: Avoids requesting the geolocation permission on page load +1ms
  LH:status Auditing: No issues in the `Issues` panel in Chrome Devtools +0ms
  LH:status Auditing: Avoids `document.write()` +0ms
  LH:status Auditing: Detected JavaScript libraries +0ms
  LH:status Auditing: Avoids requesting the notification permission on page load +0ms
  LH:status Auditing: Allows users to paste into input fields +0ms
  LH:status Auditing: Use HTTP/2 +0ms
  LH:status Auditing: Uses passive listeners to improve scrolling performance +1ms
  LH:status Auditing: Document has a meta description +0ms
  LH:status Auditing: Page has successful HTTP status code +0ms
  LH:status Auditing: Document uses legible font sizes +0ms
  LH:status Auditing: Links have descriptive text +2ms
  LH:status Auditing: Links are crawlable +1ms
  LH:status Auditing: Page isn’t blocked from indexing +0ms
  LH:status Auditing: robots.txt is valid +1ms
  LH:status Auditing: Tap targets are sized appropriately +0ms
  LH:status Auditing: Document has a valid `hreflang` +0ms
  LH:status Auditing: Document avoids plugins +1ms
  LH:status Auditing: Document has a valid `rel=canonical` +0ms
  LH:status Auditing: Structured data is valid +0ms
  LH:status Auditing: Page didn't prevent back/forward cache restoration +0ms
  LH:status Generating results... +0ms
  LH:Printer json output written to ./test.json +11ms
  LH:ChromeLauncher Killing Chrome instance 74957 +0ms
```

### PR DESCRIPTION
**Pull request by alexnj**
Fixes #14824. More details on the issue.

- Fixes the issue that resulted in full height of the page being clipped in some cases.
- Fixes discrepancy in dimensions due to horizontal overflow (overflow-x) of elements on the page.
- Simplifies the calculation a bit by relying solely on DevTools metrics API.

### CODE REVIEW

**Review by adamraine — COMMENTED** [review]
I'm going to play around with this, but can you add a smoke test that tests the horizontal overflow case.

File: core/test/gather/gatherers/full-page-screenshot-test.js [hunk 1]
```diff
@@ -11,24 +11,21 @@ import FullPageScreenshotGatherer from '../../../gather/gatherers/full-page-scre
...
-  screenshotAreaSize = contentSize;
   screenshotData = [];
   mockContext = createMockContext();
   mockContext.driver.defaultSession.sendCommand.mockImplementation((method) => {
     if (method === 'Page.getLayoutMetrics') {
       return {
         cssContentSize: contentSize,
         // See comment within _takeScreenshot() implementation
-        cssLayoutViewport: {clientWidth: screenSize.width, clientHeight: screenSize.height},
+        cssVisualViewport: {clientWidth: contentSize.width, clientHeight: contentSize.height},
```
Comment (adamraine) [inline core/test/gather/gatherers/full-page-screenshot-test.js:28]: Honestly I think we should just remove this unit test file. It requires us to effectively simulate how chrome handles all these different width/height scales. We're better off using smoke tests for this gatherer IMO.

File: core/gather/gatherers/full-page-screenshot.js [hunk 2]
```diff
@@ -103,7 +93,7 @@ class FullPageScreenshot extends FRGatherer {
       mobile: deviceMetrics.mobile,
       deviceScaleFactor: 1,
       height,
-      width: 0, // Leave width unchanged
+      width,
```
Comment (adamraine) [inline core/gather/gatherers/full-page-screenshot.js:96]: As far as I can tell, the only benefit we get from this change is capping the width at the max webp size but that will only ever be the case if we emulate a device at that extreme size.

I would prefer to just leave `width: 0` so we affect this as little as possible.

**Comment by adamraine** [comment]
Interesting I'm getting a report error on this page:
```html
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=0.5">
  <title>Document</title>
</head>
<body style="overflow: scroll;">
  <h1>AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA</h1>
</body>
</html>
```

File: core/gather/gatherers/full-page-screenshot.js (hunk 2 above)
Comment (alexnj) [reply core/gather/gatherers/full-page-screenshot.js]: Seems like it has no other side effect on calculations. Reverted to `width: 0` that might support Flow tests better.

File: core/gather/gatherers/full-page-screenshot.js [hunk 3]
```diff
@@ -128,18 +117,14 @@ class FullPageScreenshot extends FRGatherer {
...
-      });
-
     return {
       data,
-      width: screenshotAreaSize.width,
-      height: screenshotAreaSize.height,
+      // Since we resized emulated viewport to match the desired screenshot size,
+      // it is safe to rely on scaled visual viewport css dimensions.
+      width: Math.round(metrics.cssVisualViewport.clientWidth * metrics.cssVisualViewport.scale),
+      height: Math.round(metrics.cssVisualViewport.clientHeight * metrics.cssVisualViewport.scale),
```
[the remark this answers is not in the document]
Comment (connorjclark) [reply core/gather/gatherers/full-page-screenshot.js:127]: Ah, it should be exactly the same I guess. Seems good.

**Comment by connorjclark** [comment]
Tested on a handful of sites, looks good to me!

I did notice some element screenshots that were _barely_ off on cnn.com 
image

which makes me suspect the rounding...

**Comment by alexnj** [comment]
@connorjclark that looks very similar to a macOS scrollbar-width as well. Are you using an external mouse/any settings from scrollbar defaults has been changed? It shouldn't affect the left coordinates but I wonder if the horizontal translation is getting affected by it.

**Comment by connorjclark** [comment]
I believe on Mac the default is a overlayed scrollbar (takes no width). I changed that to a fixed scrollbar, because function>form

File: cli/test/smokehouse/test-definitions/fps-scaled.js [hunk 4]
```diff
@@ -46,8 +46,8 @@ const expectations = {
         ],
       },
       screenshot: {
-        height: 2000,
-        width: 824,
```
Comment (adamraine) [inline cli/test/smokehouse/test-definitions/fps-scaled.js:50]: The bounding boxes for elements do not line up with these image dimensions. I think the previous expectations for this smoke test were correct:

Screenshot 2023-04-18 at 3 01 14 PM

**Review by adamraine — CHANGES_REQUESTED** [review]

**Comment by alexnj** [comment]
@connorjclark so the reason for the minor mismatch in lining up horizontal coordinates was macOS scrollbar setting, depending on if you set `auto`/`always`/`while-scrolling`. Same issue that @adamraine found earlier as well. I don't know if there's a reliable way to test this in automation. 

The scenarios that we need to cover are (that I tested manually today):
 - device dpi differences
 - zoom/scaling factor
 - horizontal scroll
 - on desktop, specifically macOS: scrollbar setting (always vs. while scrolling)

The `fps-*` series of smoke tests cover the first two.

File: cli/test/smokehouse/test-definitions/fps-scaled.js (hunk 4 above)
Comment (adamraine) [reply cli/test/smokehouse/test-definitions/fps-scaled.js:50]: @alexnj I'm still seeing this issue and I think the old expectations were correct. Try adding an a11y failure to the `<h1>` element (e.g. make it `<h1 role="adsf'>`) and then you will see this in the report.

**Review by adamraine — APPROVED** [review]
Approving to remove my request changes blocker but someone else should review.

After testing this myself, I think our calculation for the desired emulated height was correct. However we were using the incorrect metrics for the final screenshot size.
